File: C/sorting/insertionsort.c
#include <stdio.h>
#include <stdlib.h>

/*
 * Implementation of the Insertion Sort
 */

void insertion_sort(int [], int);
void print_array(int [], int);

int
main(int argc, char **argv) {
	int array[20] = {
		 9, 71,  2, 84, 12, 23, 99, 26,  1, 76,
		67,  4, 53, 42, 10, 38, 66,  7, 80, 2
	};

	print_array(array, 20);
	insertion_sort(array, 20);
	print_array(array, 20);

	return EXIT_SUCCESS;
};

/*
 * Print the array
 */
void
print_array(int array[], int length) {
	int i;
	for (i = 0; i < length; i++)
		printf("%3d ", array[i]);
	printf("\n");
}

/*
 * Sort the (integer) array with insertion sort
 */
void
insertion_sort(int array[], int length) {
	int i, j, val;

	for (i = 1; i < length; i++) {
		val = array[i];

		j = i - 1;
		/* take the next item into the unsorted part (val) and move it
		 * to the sorted one in the right position by shifting sorted
		 * elements
		 */
		while (j >= 0 && array[j] > val) {
			array[j + 1] = array[j];
			j--;
		}

		array[j + 1] = val;
	}
}

File: java/sorting/InsertionSort.java
package sorting.java;

import java.util.Arrays;

/**
 * Simple implementation of insertion sort in java
 */
public class InsertionSort {
  public static <T extends Comparable<T>> T[] insertionsort(T[] items) {
    int size = items.length;
    int i, j;

    for (i = 1; i < size; i++) {
      j = i;

      while (j > 0 && (items[j].compareTo(items[j - 1]) < 0)) {
        swap(items, j, j - 1);
        j--;
      }
    }

    return items;
  }

  private static <T extends Comparable<T>> void swap(T[] items, int x, int y) {
    T tmp;
    tmp = items[x];
    items[x] = items[y];
    items[y] = tmp;
  }

  public static void main(String[] args) {
    System.out.println(Arrays.toString(insertionsort(new Integer[]{1, 2, 3, 4, 5, 6, 7})));
    System.out.println(Arrays.toString(insertionsort(new Integer[]{5, 1, 6, 3, 4, 2, 7})));
    System.out.println(Arrays.toString(insertionsort(new Integer[]{7, 6, 5, 4, 3, 2, 1})));

    System.out.println(Arrays.toString(insertionsort(new String[]{"a", "b", "c", "d", "e"})));
    System.out.println(Arrays.toString(insertionsort(new String[]{"b", "c", "a", "e", "d"})));
    System.out.println(Arrays.toString(insertionsort(new String[]{"e", "d", "c", "b", "a"})));
  }
}